### PR TITLE
feat: metrics middleware improvements and add logging middleware

### PR DESCRIPTION
**Pull request by zac-li**
integration tests 💚 : https://github.com/jina-ai/langchain-serve/actions/runs/5041759107

update:
https://github.com/jina-ai/langchain-serve/actions/runs/5053871547

### CODE REVIEW

File: lcserve/backend/gateway.py [hunk 1]
```diff
@@ -1067,32 +1077,109 @@ def __init__(
...
+            async def custom_send(message: dict) -> None:
+                nonlocal is_accepted
+                nonlocal status_code
+
+                if request_id and message.get('type') == 'http.response.start':
+                    message.setdefault('headers', []).append(
+                        (b'X-API-Request-ID', str(request_id).encode())
+                    )
+                    status_code = message.get('status')
+                    print(status_code)
```
Comment (deepankarm) [inline lcserve/backend/gateway.py:1150]: Left over?

File: lcserve/backend/gateway.py [hunk 2]
```diff
@@ -1067,32 +1077,109 @@ def __init__(
...
+                if is_accepted and message.get('type') not in [
+                    'websocket.send',
+                    'websocket.close',
+                ]:
+                    await original_send(
+                        {
+                            "type": "websocket.send",
+                            "text": f"connection_id:{connection_id}",
+                        }
+                    )
```
Comment (deepankarm) [inline lcserve/backend/gateway.py:1167]: Is this the first message we send to the client? This is a breaking change, not sure we can afford it now.

File: tests/integration/jcloud/test_basic_app.py [hunk 3]
```diff
@@ -38,11 +38,11 @@ async def _test_ws_route(app_id):
         await websocket.send(json.dumps({"interval": 1}))
 
         received_messages = []
-        for _ in range(5):
+        for _ in range(6):
             message = await websocket.recv()
             received_messages.append(message)
 
-        assert received_messages == ["0", "1", "2", "3", "4"]
+        assert received_messages[1:] == ["0", "1", "2", "3", "4"]
```
Comment (deepankarm) [inline tests/integration/jcloud/test_basic_app.py:45]: Is this because the first message is the connection id?

**Review by deepankarm — CHANGES_REQUESTED** [review]

File: lcserve/backend/gateway.py (hunk 1 above)
Comment (zac-li) [reply lcserve/backend/gateway.py:1150]: oops yea

File: tests/integration/jcloud/test_basic_app.py (hunk 3 above)
Comment (zac-li) [reply tests/integration/jcloud/test_basic_app.py:45]: yea

File: lcserve/backend/gateway.py (hunk 2 above)
Comment (zac-li) [reply lcserve/backend/gateway.py:1167]: dm-ed

**Review by deepankarm — APPROVED** [review]

**Pull request merged**